### PR TITLE
Support fragments in <Switch />

### PR DESCRIPTION
**Pull request by bripkens**
When checking for matched routes in `<Switch />`, traverse fragments to
determine the matched route.

Fixes https://github.com/ReactTraining/react-router/issues/5785

### CODE REVIEW

File: packages/react-router/modules/__tests__/Switch-test.js [hunk 1]
```diff
@@ -282,6 +282,67 @@ describe("A <Switch>", () => {
...
+          <Route path="/three" render={() => <h1>three</h1>} />
+        </Switch>
+      </MemoryRouter>,
+      node
+    );
+
+    expect(node.innerHTML).toMatch(/two/);
+  });
+
+  it("does handle nested fragments", () => {
```
Comment (timdorr) [inline packages/react-router/modules/__tests__/Switch-test.js]: We specifically should not support this. We don't support nested routes in a Switch, so this behavior difference will undoubtedly cause confusion. To be clear, I see this:
```js

<Switch>
  <Route path="/foo" component={Foo} />
  <>
    <Route path="/bar" component={Bar} />
    <Route path="/baz" component={Baz} />
  </>
</Switch>
```
as equivalent to this:
```js
<Switch>
  <Route path="/foo" component={Foo} />
  {[barRoute, bazRoute].map(route => <Route path={route.path} component={route.component} />)}
</Switch>
````
Fragments are basically convenience functions for arrays. So, I'd treat them as such to maintain the status quo with Switch's behaviors.

File: packages/react-router/modules/Switch.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import React from "react";
+import React, { Fragment } from "react";
```
[the remark this answers is not in the document]
Comment (bripkens) [reply packages/react-router/modules/Switch.js]: Fair point. Changed.

File: packages/react-router/modules/Switch.js [hunk 3]
```diff
@@ -42,10 +42,28 @@ class Switch extends React.Component {
...
-    let match, child;
-    React.Children.forEach(children, element => {
-      if (match == null && React.isValidElement(element)) {
+    return match
+      ? React.cloneElement(child, { location, computedMatch: match })
+      : null;
+  }
+}
+
+function getMatch(route, location, children) {
```
Comment (timdorr) [inline packages/react-router/modules/Switch.js]: Can we inline this? We don't re-use this function, so it's just indirection.

Comment (timdorr) [inline packages/react-router/modules/Switch.js]: Can we move this within the component? We don't need it externally and then we can access the props/context directly.

File: packages/react-router/modules/__tests__/Switch-test.js (hunk 1 above)
Comment (timdorr) [reply packages/react-router/modules/__tests__/Switch-test.js]: OK, I'm convinced. I didn't know about that behavior. Thanks for explaining!

File: packages/react-router/modules/Switch.js (hunk 3 above)
Comment (bripkens) [reply packages/react-router/modules/Switch.js]: I can move it within the component sure, but do know that this was a deliberate decision for the sake of efficiency. There is no need to redefine it over and over again within `render`. Especially considering that it does not need access to `this`, `props` or `state`.

Please let me know whether you still would like this to be inlined.

File: packages/react-router/modules/__tests__/Switch-test.js (hunk 1 above)
Comment (timdorr) [reply packages/react-router/modules/__tests__/Switch-test.js]: Actually, I'm not. Sorry, just busted out some code here and contradicted myself. Apologies for the flip flop!
```js
React.Children.count([<div/>, <div/>, [<div />, <div/>]]) == 4
React.Children.count([
  <div/>,
  <div/>,
  <React.Fragment>
    <div/>
    <div/>
  </React.Fragment>
]) == 3
```
Fragments are not the same as arrays, they just have a similar use case. The post here is clarifying: https://github.com/facebook/react/issues/11859#issuecomment-355654128

I'd much rather do `React.Children.map(children, child => child.type === React.Fragment ? child.props.children : child).forEach(child => ...)`. It's not common to have nested child arrays inside of `<Switch>`, so trying to emulate that behavior with something that's intentionally inconsistent.

Comment (bripkens) [reply packages/react-router/modules/__tests__/Switch-test.js]: I do not see why fragments should be handled differently for all intents and purposes of react-router than arrays. They are element containers and should be handled accordingly. 

Design decisions have been made for fragments which result in different behavior for `React.Children.count`, `React.Children.map` and others when comparing fragment and array behavior. This difference in behavior stems from important differences when handling these containers as part of `render`. React-router doesn't have to behave different when encountering fragments vs. arrays.

Furthermore, it makes no sense for react-router to try to `matchPath` a fragment element. This will never work.

Lastly, there is no technical reason why fragment recursion cannot be supported. There is no additional cost associated to it. It is faster than a `map` followed by a `forEach`. It makes array/fragment behavior consistent.

Comment (timdorr) [reply packages/react-router/modules/__tests__/Switch-test.js]: > I do not see why fragments should be handled differently for all intents and purposes of react-router than arrays. 

Because they're not arrays. I'd really rather not try to go against the upstream behavior semantics, even if it is admittedly weird. We got burned in previous versions by trying to do too much of React's behavior on our own. The edge cases are death by a thousand cuts.

I'm still on Team No Flatten™, but the rules dictate we have two maintainers approve before merging. If anyone else wants to voice their opinions, even if it's different from my own, they can and we can work towards a resolution.

Comment (jmeas) [reply packages/react-router/modules/__tests__/Switch-test.js]: I'm not a maintainer, but I recommended this change over in #5785 , so I thought I'd voice that I'm also on Team No Flatten for whatever that's worth. Thanks for all of your work @timdorr and @bripkens!

File: packages/react-router/modules/Switch.js (hunk 3 above)
Comment (jmeas) [reply packages/react-router/modules/Switch.js]: @bripkens moving it within the component doesn't necessarily mean defining it within render, right? It could go on the instance, or possibly as a static class method if React Router transpiles those down.

Comment (bripkens) [reply packages/react-router/modules/Switch.js]: Whoops, I thought the component was just a function when I looked at it again on GitHub. Sorry for that! Changed it accordingly.

File: packages/react-router/modules/__tests__/Switch-test.js (hunk 1 above)
Comment (bripkens) [reply packages/react-router/modules/__tests__/Switch-test.js]: Added the requested example to the issue:

https://github.com/ReactTraining/react-router/issues/5785#issuecomment-361852111

**Comment by mjackson** [comment]
What's your use case, @bripkens? I can't really think of a compelling use case for having a `<Fragment>` inside of a `<Switch>`.

The only thing I can imagine that you might want to do is return multiple `<Route>`s from one of your own components, e.g. a `<UserRoutes>` component that returns a `<Fragment>` containing multiple `<Route>`s. But that's something that we don't currently support.

The lack of a clear use case, plus the fact that React treats arrays differently from `<Fragment>`s is enough for me to think this isn't a good idea.

**Comment by jamesplease** [comment]
@mjackson i listed a use case over in https://github.com/ReactTraining/react-router/issues/5785#issuecomment-357286789

tl;dr keys are unnecessary boilerplate with dynamically-generated routes. 

> plus the fact that React treats array differently from `<Fragment/>`

🤷‍♂️ the users of React Router are not going to care about this difference. They are just going to want to use Fragments so that they don’t have to specify keys when mapping routes.
